tests/python/pants_test/backend/codegen/tasks/test_simple_codegen_task.py

```diff
@@ -0,0 +1,185 @@
+      'project/src/main/genfoo/foo-bar:foo-target-a',
+    ]
+    spec_paths = set([spec.split(':')[0] for spec in target_specs])
+    spec_names = set([spec.split(':')[1] for spec in target_specs])
+    for spec_path in spec_paths:
+      parent_dir = os.path.dirname(spec_path)
+      if not os.path.exists(parent_dir):
+        self.create_dir(parent_dir)
+      self.create_file('{0}/BUILD'.format(spec_path), '\n'.join([dedent('''
+          java_protobuf_library(name='{name}',
+            sources=[],
+            dependencies=[],
+          )
+        '''.format(name=spec_name)) for spec_name in spec_names]))
+
+    targets = set([self.target(spec) for spec in target_specs])
+
+    get_suffix = SimpleCodegenTask._codegen_workdir_suffix
+
+    for target in targets:
+      self.assertEqual('global', get_suffix(target, 'global'))
+      self.assertTrue('isolated' in get_suffix(target, 'isolated'))
+
+    global_dirs = set(get_suffix(target, 'global') for target in targets)
+    isolated_dirs = set(get_suffix(target, 'isolated') for target in targets)
+
+    self.assertEqual(1, len(global_dirs), 'There should only be one global directory suffix!')
+    self.assertEqual(len(targets), len(isolated_dirs),
+                     'There should be exactly one directory suffix per unique target!')
+
+    for target in targets:
+      for strategy in ('global', 'isolated',):
+        suffix = None
+        for i in range(5):
+          curr_suffix = get_suffix(target, strategy)
+          if i > 0:
+            self.assertEqual(suffix, curr_suffix,
+                             'Suffix should not be unstable for the same target and strategy! '
+                             '({target} with {strategy} got {suffix_a} and {suffix_b})'
+                             .format(target=target.id,
+                                     strategy=strategy,
+                                     suffix_a=suffix,
+                                     suffix_b=curr_suffix))
+          suffix = curr_suffix
+
+  def test_execute(self):
+    proto_suffixes = ['a', 'b', 'c',]
+
+    self.add_to_build_file('gen-lib', '\n'.join(dedent('''
+      java_wire_library(name='{suffix}',
+        sources=['org/pantsbuild/example/foo{suffix}.proto'],
+      )
+    ''').format(suffix=suffix) for suffix in proto_suffixes))
+
+    for suffix in proto_suffixes:
+      self.create_file('gen-lib/org/pantsbuild/example/foo{suffix}.proto'.format(suffix=suffix),
+                       dedent('''
+                                package org.pantsbuild.example;
+
+                                message Foo%s {
+                                  optional string bar = 1;
+                                }
+                              ''' % suffix)) # .format() chokes because of the message braces.
+
+    targets = [self.target('gen-lib:{suffix}'.format(suffix=suffix)) for suffix in proto_suffixes]
+
+    test_base = self
+    class TestGen(SimpleCodegenTask):
```

Comment:
It might be worth pulling this out of this single test case.  It could be useful for another test, and you are having to shadow self from the test method.